services/src/main/java/org/keycloak/protocol/oidc/AccessTokenIntrospectionProvider.java

```diff
@@ -97,6 +111,49 @@ public Response introspect(String token) {
         }
     }
 
+    private AccessToken getAccessToken(String token) throws VerificationException {
+        TokenVerifier<AccessToken> verifier = TokenVerifier.create(token, AccessToken.class);
+        SignatureVerifierContext verifierContext = session.getProvider(SignatureProvider.class, verifier.getHeader().getAlgorithm().name()).verifier(verifier.getHeader().getKeyId());
+        verifier.verifierContext(verifierContext);
+        AccessToken accessToken = verifier.verify().getToken();
+        SingleUseObjectProvider singleUseStore = session.singleUseObjects();
+        Map<String, String> tokenData = singleUseStore.get(accessToken.getId());
+        if (tokenData != null) {
+            // enabled lightWeightAccessTokenMapper
+            accessToken.setSessionState(tokenData.get(SESSION_ID));
+            accessToken.issuedFor(tokenData.get(AZP));
+            return accessToken;
+        } else {
+            verifier.realmUrl(Urls.realmIssuer(session.getContext().getUri().getBaseUri(), realm.getName()));
+            RealmModel realm = this.session.getContext().getRealm();
+            return tokenManager.checkTokenValidForIntrospection(session, realm, accessToken, false) ? accessToken : null;
+        }
+    }
+
+
+    private AccessToken transformAccessToken(AccessToken token) {
+        if (token == null) {
+            return null;
+        }
+
+        ClientModel client = realm.getClientByClientId(token.getIssuedFor());
+        UserSessionModel userSession = new UserSessionCrossDCManager(session).getUserSessionWithClient(realm, token.getSessionId(), false, client.getId());
+        AuthenticatedClientSessionModel clientSession = userSession.getAuthenticatedClientSessionByClient(client.getId());
```

Comment:
This does not take into account special case access tokens ( token.getSessionState() == null - fe service account, offline access tokens).

Please see methods _TokenManager.checkTokenValidForIntrospection_ and _UserInfoEndpoint.findValidSession_.

Reply:
UserInfoEndpoint.findValidSession has been Utilized (UserSessionUtil) so that it can also be referenced by the AccessTokenIntrospectionProvider.

Reply:
> UserInfoEndpoint.findValidSession has been Utilized (UserSessionUtil) so that it can also be referenced by the AccessTokenIntrospectionProvider.

Now it is correct. I only believe that we should catch method errors and together with user being null return null in this method.
So token introspection will return failed - not exception.